plugin/backend/ipvariant.d

```diff
@@ -0,0 +1,906 @@
+        import std.typecons : scoped;
+        import cpptooling.analyzer.clang.analyze_helper : ClassVisitor;
+        import cpptooling.analyzer.clang.type : retrieveType;
+        import cpptooling.analyzer.clang.utility : put;
+
+        ///TODO add information if it is a public/protected/private class.
+        ///TODO add metadata to the class if it is a definition or declaration
+
+        mixin(mixinNodeLog!());
+        logger.info("class: ", v.cursor.spelling);
+
+        if (v.cursor.isDefinition) {
+            auto visitor = scoped!ClassVisitor(v, ns_stack, container, indent + 1);
+            v.accept(visitor);
+
+            root.put(visitor.root);
+            container.put(visitor.root, visitor.root.fullyQualifiedName);
+        } else {
+            auto type = retrieveType(v.cursor, container, indent);
+            put(type, container, indent);
+        }
+    }
+
+    override void visit(const(Namespace) v) @trusted {
+        mixin(mixinNodeLog!());
+
+        () @trusted{ ns_stack ~= CppNs(v.cursor.spelling); }();
+        // pop the stack when done
+        scope (exit)
+            ns_stack = ns_stack[0 .. $ - 1];
+
+        auto ns_visitor = scoped!(CppVisitor!(CppNamespace, ControllerT, ProductT))(ctrl,
+                prod, indent, ns_stack, container);
+
+        // fill the namespace with content from the analyse
+        v.accept(ns_visitor);
+        root.put(ns_visitor.root);
+    }
+
+    override void visit(const(TranslationUnit) v) {
+        import std.algorithm : filter;
+        import cpptooling.analyzer.clang.type : makeLocation;
+
+        mixin(mixinNodeLog!());
+
+        LocationTag tu_loc;
+        () @trusted{ tu_loc = LocationTag(Location(v.cursor.spelling, 0, 0)); }();
+
+        if (tu_loc.kind != LocationTag.Kind.noloc && ctrl.doFile(tu_loc.file,
+                "root " ~ tu_loc.toString)) {
+            prod.putLocation(FileName(tu_loc.file), LocationType.Root);
+        }
+
+        v.accept(this);
+    }
+
+    void toString(Writer)(scope Writer w) @safe const {
+        import std.format : FormatSpec;
+        import std.range.primitives : put;
+
+        auto fmt = FormatSpec!char("%u");
+        fmt.writeUpToNextSpec(w);
+
+        root.toString(w, fmt);
+        put(w, "\n");
+        container.get.toString(w, FormatSpec!char("%s"));
+    }
+
+    override string toString() const {
+        import std.exception : assumeUnique;
+
+        char[] buf;
+        buf.reserve(100);
+        toString((const(char)[] s) { buf ~= s; });
+        auto trustedUnique(T)(T t) @trusted {
+            return assumeUnique(t);
+        }
+
+        return trustedUnique(buf);
+    }
+}
+
+private:
+@safe:
+
+import cpptooling.data.representation : CppRoot, CppClass, CppMethod, CppCtor,
+    CppDtor, CFunction, CppNamespace;
+import cpptooling.data.type : LocationTag, Location;
+import cpptooling.data.symbol.container : Container;
+import dsrcgen.cpp : E;
+
+enum ClassType {
```

Comment:
Adapt ClassType for _your needs_. Gmock is Google Mock. I don't think anything is generated that is Google Mock.